### PR TITLE
handle protocol.json parse exception

### LINKED ISSUE #1412: [NEO3] there is not JSON Format error show when protocol.json is with json format error
**Describe the bug**
when start a neo-cli, I delete a "}" in the last line of the protocol.json by mistake, and unaware of it. then start neo-cli, it shows "Object reference not set to an instance of an object" error, with all the error messages, i can not get the information of the json error.
If config.json is with format error, there shows "Can not parse json", which is expected.

**To Reproduce**
neo-cli: master 51cd29fbe21abb9e1f17f64e5c6d21bc7decbbb9
neo: master ab4830c
neo-vm: master be2ac36bf35a3033d828e0ba0630d390599c487d
MillisecondsPerBlock is 15000.
1. Start a single consensus node.
2. Start a external node.
3. modify the protocol.json, such as delete a "}" in the last line.
```
{
  "ProtocolConfiguration": {
    "Magic": 20200109,
    "MillisecondsPerBlock": 15000,
    "StandbyValidators": [
      "03ac37bdaa57563af356a787da8503db3bdfbde1f072c3bc3a6433b1929e190e4d"
    ],
    "SeedList": [
      "localhost:13333"
    ]
  }
```
4. start neo-cli, it shows:
```
neo> Unhandled exception. System.NullReferenceException: Object reference not set to an instance of an object.
   at Neo.Plugins.Logger.Neo.Plugins.ILogPlugin.Log(String source, LogLevel level, String message)
   at Neo.Utility.Log(String source, LogLevel level, String message) in D:\NEO\Github\neo\src\neo\Utility.cs:line 25
   at Neo.Plugins.Plugin.LoadPlugin(Assembly assembly) in D:\NEO\Github\neo\src\neo\Plugins\Plugin.cs:line 134
   at Neo.Plugins.Plugin.LoadPlugins(NeoSystem system) in D:\NEO\Github\neo\src\neo\Plugins\Plugin.cs:line 154
   at Neo.NeoSystem..ctor(String storageEngine) in D:\NEO\Github\neo\src\neo\NeoSystem.cs:line 32
   at Neo.CLI.MainService.Start(String[] args)
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__139_1(Object state)
   at System.Threading.QueueUserWorkItemCallbackDefaultContext.Execute()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()
```

**Expected behavior**
here is error message of neo-cli 2.10.3 for reference.
039650e3bd6792038cf906bfa562782

7081bd82be1329adeaf741aabaa6b19

**Screenshots**
ee2798aca4644e2f446ee547b3794b7
e620bbb5d4f4037f78b2e279e8e948b

**Platform:**
 - OS: Windows 10 x64

### PR DESCRIPTION
**Pull request by KickSeason**
fix #1412 
@cloud8little  Try this.
Plugin will use  protocol settings in their constructor.

### CODE REVIEW

**Comment by cloud8little** [comment]
Test cases:

- First User Scenario  [PASS]
1. Start a single CN node
2. Start a external node.
3. modify the protocol.json, such as delete a "}" in the last line.

```
{
  "ProtocolConfiguration": {
    "Magic": 20200109,
    "MillisecondsPerBlock": 15000,
    "StandbyValidators": [
      "03ac37bdaa57563af356a787da8503db3bdfbde1f072c3bc3a6433b1929e190e4d"
    ],
    "SeedList": [
      "localhost:13333"
    ]
  }
```

Plugins: 
1. only LevelDBStore
2. LevelDBStore + SystemLog
3. LevelDBStore + SystemLog + RpcServer
4. LevelDBStore + SystemLog + RpcServer + RpcNep5Tracker + ApplicationLogs
5. No plugins

All situation print following error message.
```
D:\NEO\NEO3\privatenet\test-neo-pr1426\external>dotnet neo-cli.dll
Failed parsing protocol.json. Using default settings
System.FormatException: Could not parse the JSON file.
 ---> System.Text.Json.JsonReaderException: Expected depth to be zero at the end of the JSON payload. There is an open JSON object or array that should be closed. LineNumber: 12 | BytePositionInLine: 0.
   at System.Text.Json.ThrowHelper.ThrowJsonReaderException(Utf8JsonReader& json, ExceptionResource resource, Byte nextByte, ReadOnlySpan`1 bytes)
   at System.Text.Json.Utf8JsonReader.ReadSingleSegment()
   at System.Text.Json.Utf8JsonReader.Read()
   at System.Text.Json.JsonDocument.Parse(ReadOnlySpan`1 utf8JsonSpan, Utf8JsonReader reader, MetadataDb& database, StackRowStack& stack)
   at System.Text.Json.JsonDocument.Parse(ReadOnlyMemory`1 utf8Json, JsonReaderOptions readerOptions, Byte[] extraRentedBytes)
   at System.Text.Json.JsonDocument.Parse(ReadOnlyMemory`1 json, JsonDocumentOptions options)
   at System.Text.Json.JsonDocument.Parse(String json, JsonDocumentOptions options)
   at Microsoft.Extensions.Configuration.Json.JsonConfigurationFileParser.ParseStream(Stream input)
   at Microsoft.Extensions.Configuration.Json.JsonConfigurationFileParser.Parse(Stream input)
   at Microsoft.Extensions.Configuration.Json.JsonConfigurationProvider.Load(Stream stream)
   --- End of inner exception stack trace ---
   at Microsoft.Extensions.Configuration.Json.JsonConfigurationProvider.Load(Stream stream)
   at Microsoft.Extensions.Configuration.FileConfigurationProvider.Load(Boolean reload)
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Extensions.Configuration.FileConfigurationProvider.HandleException(ExceptionDispatchInfo info)
   at Microsoft.Extensions.Configuration.FileConfigurationProvider.Load(Boolean reload)
   at Microsoft.Extensions.Configuration.FileConfigurationProvider.Load()
   at Microsoft.Extensions.Configuration.ConfigurationRoot..ctor(IList`1 providers)
   at Microsoft.Extensions.Configuration.ConfigurationBuilder.Build()
   at Neo.Utility.LoadConfig(String config) in D:\NEO\Github\neo\src\neo\Utility.cs:line 18
   at Neo.ProtocolSettings.get_Default() in D:\NEO\Github\neo\src\neo\ProtocolSettings.cs:line 40
```

- Second User Scenario   [PASS] Note: there is also "Operation is not valid due to the current state of the object." error message print after JSON parse error. From my view, it is OK. 
1. Start a single CN node
2. modify the protocol.json, such as delete a "}" in the last line.

```
{
  "ProtocolConfiguration": {
    "Magic": 20200109,
    "MillisecondsPerBlock": 15000,
    "StandbyValidators": [
      "03ac37bdaa57563af356a787da8503db3bdfbde1f072c3bc3a6433b1929e190e4d"
    ],
    "SeedList": [
      "localhost:13333"
    ]
  }
```
3. Start CN node.

```
D:\NEO\NEO3\privatenet\test-neo-pr1426\cns>dotnet neo-cli.dll
Failed parsing protocol.json. Using default settings
System.FormatException: Could not parse the JSON file.
 ---> System.Text.Json.JsonReaderException: Expected depth to be zero at the end of the JSON payload. There is an open JSON object or array that should be closed. LineNumber: 12 | BytePositionInLine: 0.
   at System.Text.Json.ThrowHelper.ThrowJsonReaderException(Utf8JsonReader& json, ExceptionResource resource, Byte nextByte, ReadOnlySpan`1 bytes)
   at System.Text.Json.Utf8JsonReader.ReadSingleSegment()
   at System.Text.Json.Utf8JsonReader.Read()
   at System.Text.Json.JsonDocument.Parse(ReadOnlySpan`1 utf8JsonSpan, Utf8JsonReader reader, MetadataDb& database, StackRowStack& stack)
   at System.Text.Json.JsonDocument.Parse(ReadOnlyMemory`1 utf8Json, JsonReaderOptions readerOptions, Byte[] extraRentedBytes)
   at System.Text.Json.JsonDocument.Parse(ReadOnlyMemory`1 json, JsonDocumentOptions options)
   at System.Text.Json.JsonDocument.Parse(String json, JsonDocumentOptions options)
   at Microsoft.Extensions.Configuration.Json.JsonConfigurationFileParser.ParseStream(Stream input)
   at Microsoft.Extensions.Configuration.Json.JsonConfigurationFileParser.Parse(Stream input)
   at Microsoft.Extensions.Configuration.Json.JsonConfigurationProvider.Load(Stream stream)
   --- End of inner exception stack trace ---
   at Microsoft.Extensions.Configuration.Json.JsonConfigurationProvider.Load(Stream stream)
   at Microsoft.Extensions.Configuration.FileConfigurationProvider.Load(Boolean reload)
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Extensions.Configuration.FileConfigurationProvider.HandleException(ExceptionDispatchInfo info)
   at Microsoft.Extensions.Configuration.FileConfigurationProvider.Load(Boolean reload)
   at Microsoft.Extensions.Configuration.FileConfigurationProvider.Load()
   at Microsoft.Extensions.Configuration.ConfigurationRoot..ctor(IList`1 providers)
   at Microsoft.Extensions.Configuration.ConfigurationBuilder.Build()
   at Neo.Utility.LoadConfig(String config) in D:\NEO\Github\neo\src\neo\Utility.cs:line 18
   at Neo.ProtocolSettings.get_Default() in D:\NEO\Github\neo\src\neo\ProtocolSettings.cs:line 40
NEO-CLI v3.0.0-preview1  -  NEO v3.0.0-preview1  -  NEO-VM v3.0.0-preview1

neo> [ERROR][2020/1/20 10:33:31][Thread 0010][akka://NeoSystem/user/$a] Operation is not valid due to the current state of the object.
Cause: System.InvalidOperationException: Operation is not valid due to the current state of the object.
   at Neo.Ledger.Blockchain.OnImport(IEnumerable`1 blocks, Boolean verify) in D:\NEO\Github\neo\src\neo\Ledger\Blockchain.cs:line 248
   at Akka.Actor.UntypedActor.Receive(Object message)
   at Akka.Actor.ActorBase.AroundReceive(Receive receive, Object message)
   at Akka.Actor.ActorCell.ReceiveMessage(Object message)
   at Akka.Actor.ActorCell.Invoke(Envelope envelope)
[ERROR][2020/1/20 10:33:31][Thread 0008][akka://NeoSystem/user/$a] Error while creating actor instance of type Neo.Ledger.Blockchain with 2 args: (Neo.NeoSystem,Neo.Plugins.Storage.Store)
Cause: [akka://NeoSystem/user/$a#1788993163]: Akka.Actor.PostRestartException: Exception post restart (System.InvalidOperationException)
 ---> System.TypeLoadException: Error while creating actor instance of type Neo.Ledger.Blockchain with 2 args: (Neo.NeoSystem,Neo.Plugins.Storage.Store)
 ---> System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation.
 ---> System.InvalidOperationException: Operation is not valid due to the current state of the object.
   at Neo.Ledger.Blockchain..ctor(NeoSystem system, IStore store) in D:\NEO\Github\neo\src\neo\Ledger\Blockchain.cs:line 107
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor, Boolean wrapExceptions)
   at System.Reflection.RuntimeConstructorInfo.Invoke(BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.RuntimeType.CreateInstanceImpl(BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture)
   at System.Activator.CreateInstance(Type type, BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes)
   at System.Activator.CreateInstance(Type type, Object[] args)
   at Akka.Actor.Props.ActivatorProducer.Produce()
   at Akka.Actor.Props.NewActor()
   --- End of inner exception stack trace ---
   at Akka.Actor.Props.NewActor()
   at Akka.Actor.ActorCell.CreateNewActorInstance()
   at Akka.Actor.ActorCell.<>c__DisplayClass109_0.<NewActor>b__0()
   at Akka.Actor.ActorCell.UseThreadContext(Action action)
   at Akka.Actor.ActorCell.NewActor()
   at Akka.Actor.ActorCell.FinishRecreate(Exception cause, ActorBase failedActor)
   --- End of inner exception stack trace ---
```

File: src/neo/ProtocolSettings.cs [hunk 1]
```diff
@@ -33,10 +34,19 @@ public static ProtocolSettings Default
...
-                    var configuration = Utility.LoadConfig("protocol");
+                    IConfigurationRoot configuration;
+                    try
+                    {
+                        configuration = Utility.LoadConfig("protocol");
+                    }
+                    catch (System.Exception e)
+                    {
+                        Console.WriteLine("Failed parsing protocol.json. Using default settings");
+                        Console.WriteLine(e.ToString());
```
Comment (erikzhang) [inline src/neo/ProtocolSettings.cs]: Don't use `Console.WriteLine()` in a non-console project.

Comment (Tommo-L) [reply src/neo/ProtocolSettings.cs]: @KickSeason  plz use `Utility.Log`

File: src/neo/ProtocolSettings.cs [hunk 2]
```diff
@@ -33,10 +33,19 @@ public static ProtocolSettings Default
...
-                    var configuration = Utility.LoadConfig("protocol");
+                    IConfigurationRoot configuration;
+                    try
+                    {
+                        configuration = Utility.LoadConfig("protocol");
+                    }
+                    catch (Exception e)
+                    {
+                        Utility.Log(nameof(ProtocolSettings), LogLevel.Warning, "Failed parsing protocol.json. Using default settings");
+                        Utility.Log(nameof(ProtocolSettings), LogLevel.Error, e.Message);
```
Comment (erikzhang) [inline src/neo/ProtocolSettings.cs]: Why we have two logs for a single error?

File: src/neo/ProtocolSettings.cs [hunk 3]
```diff
@@ -33,10 +33,18 @@ public static ProtocolSettings Default
...
+                    IConfigurationRoot configuration;
+                    try
+                    {
+                        configuration = Utility.LoadConfig("protocol");
+                    }
+                    catch (Exception e)
+                    {
+                        Utility.Log(nameof(ProtocolSettings), LogLevel.Error, "Failed parsing protocol.json. Using default settings. Error: " + e.Message);
+                        configuration = new ConfigurationBuilder().Build();
+                    }
```
Comment (erikzhang) [inline src/neo/ProtocolSettings.cs]: We can move the `try...catch` to `Utility.LoadConfig()`, then it can be used in other places.

Comment (KickSeason) [reply src/neo/ProtocolSettings.cs]: 👌

File: src/neo/Utility.cs [hunk 4]
```diff
@@ -15,9 +15,18 @@ public static IConfigurationRoot LoadConfig(string config)
         {
             var env = Environment.GetEnvironmentVariable("NEO_NETWORK");
             var configFile = string.IsNullOrWhiteSpace(env) ? $"{config}.json" : $"{config}.{env}.json";
-            return new ConfigurationBuilder()
+            try
+            {
+                return new ConfigurationBuilder()
                 .AddJsonFile(configFile, true)
                 .Build();
```
Comment (erikzhang) [inline src/neo/Utility.cs]: Indent or inline.

File: src/neo/Utility.cs [hunk 5]
```diff
@@ -15,9 +15,18 @@ public static IConfigurationRoot LoadConfig(string config)
...
+            }
+            catch (Exception e)
+            {
+                Log(nameof(Utility), LogLevel.Error, $"Failed parsing {configFile}, Error: " + e.Message);
+                return new ConfigurationBuilder()
+                .Build();
```
Comment (erikzhang) [inline src/neo/Utility.cs]: Indent or inline.

**Comment by vncoelho** [comment]
@KickSeason, about the format it should be something with the version.
I do not know, locally it says everything is fine about the format:

```cs
Tool 'dotnet-format' (version '3.2.107702') was successfully installed.
  Formatting code files in workspace '/home/runner/work/neo/neo/neo.sln'.
  Loading workspace.
  Msbuild failed when processing the file '/home/runner/work/neo/neo/tests/neo.UnitTests/neo.UnitTests.csproj' with message: /opt/hostedtoolcache/dncs/3.0.100/x64/sdk/3.0.100/Microsoft.Common.CurrentVersion.targets: (1540, 5): The "Microsoft.Build.Tasks.ResolveNonMSBuildProjectOutput" task could not be loaded from the assembly Microsoft.Build.Tasks.Core, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a.  Confirm that the <UsingTask> declaration is correct, that the assembly and all its dependencies are available, and that the task contains a public class that implements Microsoft.Build.Framework.ITask.
```
workflow from github

**Comment by shargon** [comment]
We have troubles with dotnet format

**Comment by KickSeason** [comment]
I got this error locally even on master branch.
image

**Comment by KickSeason** [comment]
I should say that if we use `Utility.Log` to output this exception, there will be two problems
1. Users that do not install log plugin cannot get error messages
2. Even log plugin is installed, and the exception before the log plugin is loaded will not be recorded.

**Comment by shargon** [comment]
I am not a fan of this plugin, I prefer to activate this in `neo-cli` with `--logs`, but install a plugin in order to see the logs, looks very weird to me.

**Comment by cloud8little** [comment]
@KickSeason 
with only SystemLog installed, following error message shows, which is not obvious to show protocol file format error.
8db04b3310d105295d8eed07d9a6b0f

with SystemLog + other plugins installed, no exception message shows.

File: src/neo/Plugins/Plugin.cs [hunk 6]
```diff
@@ -46,15 +46,15 @@ static Plugin()
...
+            
             Plugins.Add(this);
 
             if (this is ILogPlugin logger) Loggers.Add(logger);
             if (this is IStoragePlugin storage) Storages.Add(Name, storage);
             if (this is IP2PPlugin p2p) P2PPlugins.Add(p2p);
             if (this is IPersistencePlugin persistence) PersistencePlugins.Add(persistence);
             if (this is IMemoryPoolTxObserverPlugin txObserver) TxObserverPlugins.Add(txObserver);
-
-            Configure();
```
Comment (erikzhang) [inline src/neo/Plugins/Plugin.cs]: Any help with moving `Configure()` here?

Comment (KickSeason) [reply src/neo/Plugins/Plugin.cs]: There might be exception in when Configure. I think we should add the plugin after it configure successfully.

Comment (erikzhang) [reply src/neo/Plugins/Plugin.cs]: I prefer to remove the log in `Configure()`.

Comment (KickSeason) [reply src/neo/Plugins/Plugin.cs]: In this case users won't notice the json file format error.
What about not catching in `Utility.LoadConfig` and throwing this exception to high level.
Although the exception will be catched in `Plugin.LoadPlugin` but will be throwed when blockchain initializes. In this case Plugins load failed and Blockchain throws exception. Users can't start neo.

Comment (erikzhang) [reply src/neo/Plugins/Plugin.cs]: So this PR can be closed?

Comment (KickSeason) [reply src/neo/Plugins/Plugin.cs]: yes, I think. Merge and close.

Comment (erikzhang) [reply src/neo/Plugins/Plugin.cs]: There is nothing to be merged.

**Comment by cloud8little** [comment]
Test [FAIL]:
If With LevelDB  installed and "Engine": "LevelDBStore" written in config.json.
443e25fceb71a5dec4b7fc37b59bc7f
```
neo> Unhandled exception. System.Collections.Generic.KeyNotFoundException: The given key 'LevelDBStore' was not present in the dictionary.
   at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
   at Neo.NeoSystem..ctor(String storageEngine) in D:\NEO\Github\neo\src\neo\NeoSystem.cs:line 33
   at Neo.CLI.MainService.Start(String[] args)
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__139_1(Object state)
   at System.Threading.QueueUserWorkItemCallbackDefaultContext.Execute()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()
```

without "Engine": "LevelDBStore" written in config.json and no LevelDBStore installed:
image

**Comment by erikzhang** [comment]
Let's throw exceptions if the config file is wrong.

**Pull request closed without merging**